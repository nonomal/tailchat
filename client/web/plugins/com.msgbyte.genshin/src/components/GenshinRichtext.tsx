import React, { useMemo } from 'react';
import parser from 'html-react-parser';

/**
 * 原神富文本渲染
 */
export const GenshinRichtext: React.FC<{
  raw: string;
}> = React.memo(({ raw }) => {
  const el = useMemo(() => {
    const processedHtml = raw.replace(
      /\<color=(.*?)\>(.*?)\<\/color\>/g,
      '<span style="color: $1;">$2</span>'
    );

    return parser(processedHtml);
  }, [raw]);
  return <>{el}</>;
});
GenshinRichtext.displayName = 'GenshinRichtext';
