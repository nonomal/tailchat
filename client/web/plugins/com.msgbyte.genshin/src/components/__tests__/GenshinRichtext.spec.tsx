import { GenshinRichtext } from '../GenshinRichtext';
import { render } from '@testing-library/react';
import React from 'react';

describe('GenshinRichtext', () => {
  const testRawRichtext = `「<color=#cc9046FF>陵薮</color>市朝」活动祈愿已开启。活动期间内，<color=#c93f23>限定</color>5星角色<color=#debd6c>「尘世闲游·钟离(岩)」</color>以及4星角色<color=#00BFFF>「少年春衫薄·行秋(水)」</color>、<color=#945dc4>「无冕的龙王·北斗(雷)」</color>、<color=#EC4923>「智明无邪·烟绯(火)」</color>的祈愿获取概率将<color=#c93f23>大幅提升</color>！</P>
  <color=#c93f23>※以上角色中，限定角色不会进入「奔行世间」常驻祈愿。 </color></P>
  <br />
  ※一般情况下所有角色或武器均适用基础概率，如触发概率UP、保底等以具体规则为准。 </P>
  <br />
  〓祈愿规则〓</P>
  【5星物品】</P>
  在本期「<color=#cc9046FF>陵薮</color>市朝」活动祈愿中，5星角色祈愿的基础概率为<color=#c93f23>0.600%</color>，综合概率（含保底）为<color=#c93f23>1.600%</color>，最多<color=#c93f23>90</color>次祈愿必定能通过保底获取5星角色。</P>
  当祈愿获取到5星角色时，有<color=#c93f23>50.000%</color>的概率为本期5星UP角色<color=#debd6c>「尘世闲游·钟离(岩)」</color>。如果本次祈愿获取的5星角色非本期5星UP角色，下次祈愿获取的5星角色<color=#c93f23>必定</color>为本期5星UP角色。</P>
  【4星物品】</P>
  在本期「<color=#cc9046FF>陵薮</color>市朝」活动祈愿中，4星物品祈愿的基础概率为<color=#c93f23>5.100%</color>，4星角色祈愿的基础概率为<color=#c93f23>2.550%</color>，4星武器祈愿的基础概率为<color=#c93f23>2.550%</color>，4星物品祈愿的综合概率（含保底）为<color=#c93f23>13.000%</color>。最多<color=#c93f23>10</color>次祈愿必定能通过保底获取4星或以上物品，通过保底获取4星物品的概率为<color=#c93f23>99.400%</color>，获取5星物品的概率为<color=#c93f23>0.600%</color>。</P>
  当祈愿获取到4星物品时，有<color=#c93f23>50.000%</color>的概率为本期4星UP角色<color=#00BFFF>「少年春衫薄·行秋(水)」</color>、<color=#945dc4>「无冕的龙王·北斗(雷)」</color>、<color=#EC4923>「智明无邪·烟绯(火)」</color>中的一个。如果本次祈愿获取的4星物品非本期4星UP角色，下次祈愿获取的4星物品<color=#c93f23>必定</color>为本期4星UP角色。当祈愿获取到4星UP物品时，每个本期4星UP角色的获取概率均等。</P>
  <br />
  获得4星武器时，会同时获得2个<color=#bd6932>无主的星辉</color>作为副产物；获得3星武器时，会同时获得15个<color=#a256e1>无主的星尘</color>作为副产物。</P>
  <br />
  〓若获得重复角色〓</P>
  无论通过何种方式（包含但不限于祈愿、商城兑换、系统赠送等）第2~7次获得相同5星角色时，每次将转化为1个<color=#a256e1>对应角色的命星</color>和10个<color=#bd6932>无主的星辉</color>；第8次及之后获得，将仅转化为25个<color=#bd6932>无主的星辉</color>。</P>
  无论通过何种方式（包含但不限于祈愿、商城兑换、系统赠送等）第2~7次获得相同4星角色时，每次将转化为1个<color=#a256e1>对应角色的命星</color>和2个<color=#bd6932>无主的星辉</color>；第8次及之后获得，将仅转化为5个<color=#bd6932>无主的星辉</color>。</P>
  <br />
  ※本祈愿属于「角色活动祈愿」，「角色活动祈愿」和「角色活动祈愿-2」的祈愿次数保底完全共享，会一直共同累计在「角色活动祈愿」和「角色活动祈愿-2」中，与其他祈愿的祈愿次数保底相互独立计算，互不影响。</P>`;

  test('test', () => {
    const wrapper = render(<GenshinRichtext raw={testRawRichtext} />);
    expect(wrapper.container).toMatchSnapshot();
  });
});
