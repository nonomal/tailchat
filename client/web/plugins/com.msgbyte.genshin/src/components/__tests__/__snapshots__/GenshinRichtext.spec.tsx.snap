// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenshinRichtext test 1`] = `
<div>
  「
  <span
    style="color: rgba(204, 144, 70, 1);"
  >
    陵薮
  </span>
  市朝」活动祈愿已开启。活动期间内，
  <span
    style="color: rgb(201, 63, 35);"
  >
    限定
  </span>
  5星角色
  <span
    style="color: rgb(222, 189, 108);"
  >
    「尘世闲游·钟离(岩)」
  </span>
  以及4星角色
  <span
    style="color: rgb(0, 191, 255);"
  >
    「少年春衫薄·行秋(水)」
  </span>
  、
  <span
    style="color: rgb(148, 93, 196);"
  >
    「无冕的龙王·北斗(雷)」
  </span>
  、
  <span
    style="color: rgb(236, 73, 35);"
  >
    「智明无邪·烟绯(火)」
  </span>
  的祈愿获取概率将
  <span
    style="color: rgb(201, 63, 35);"
  >
    大幅提升
  </span>
  ！
  <p />
  
  
  <span
    style="color: rgb(201, 63, 35);"
  >
    ※以上角色中，限定角色不会进入「奔行世间」常驻祈愿。 
  </span>
  <p />
  
  
  <br />
  
  ※一般情况下所有角色或武器均适用基础概率，如触发概率UP、保底等以具体规则为准。 
  <p />
  
  
  <br />
  
  〓祈愿规则〓
  <p />
  
  【5星物品】
  <p />
  
  在本期「
  <span
    style="color: rgba(204, 144, 70, 1);"
  >
    陵薮
  </span>
  市朝」活动祈愿中，5星角色祈愿的基础概率为
  <span
    style="color: rgb(201, 63, 35);"
  >
    0.600%
  </span>
  ，综合概率（含保底）为
  <span
    style="color: rgb(201, 63, 35);"
  >
    1.600%
  </span>
  ，最多
  <span
    style="color: rgb(201, 63, 35);"
  >
    90
  </span>
  次祈愿必定能通过保底获取5星角色。
  <p />
  
  当祈愿获取到5星角色时，有
  <span
    style="color: rgb(201, 63, 35);"
  >
    50.000%
  </span>
  的概率为本期5星UP角色
  <span
    style="color: rgb(222, 189, 108);"
  >
    「尘世闲游·钟离(岩)」
  </span>
  。如果本次祈愿获取的5星角色非本期5星UP角色，下次祈愿获取的5星角色
  <span
    style="color: rgb(201, 63, 35);"
  >
    必定
  </span>
  为本期5星UP角色。
  <p />
  
  【4星物品】
  <p />
  
  在本期「
  <span
    style="color: rgba(204, 144, 70, 1);"
  >
    陵薮
  </span>
  市朝」活动祈愿中，4星物品祈愿的基础概率为
  <span
    style="color: rgb(201, 63, 35);"
  >
    5.100%
  </span>
  ，4星角色祈愿的基础概率为
  <span
    style="color: rgb(201, 63, 35);"
  >
    2.550%
  </span>
  ，4星武器祈愿的基础概率为
  <span
    style="color: rgb(201, 63, 35);"
  >
    2.550%
  </span>
  ，4星物品祈愿的综合概率（含保底）为
  <span
    style="color: rgb(201, 63, 35);"
  >
    13.000%
  </span>
  。最多
  <span
    style="color: rgb(201, 63, 35);"
  >
    10
  </span>
  次祈愿必定能通过保底获取4星或以上物品，通过保底获取4星物品的概率为
  <span
    style="color: rgb(201, 63, 35);"
  >
    99.400%
  </span>
  ，获取5星物品的概率为
  <span
    style="color: rgb(201, 63, 35);"
  >
    0.600%
  </span>
  。
  <p />
  
  当祈愿获取到4星物品时，有
  <span
    style="color: rgb(201, 63, 35);"
  >
    50.000%
  </span>
  的概率为本期4星UP角色
  <span
    style="color: rgb(0, 191, 255);"
  >
    「少年春衫薄·行秋(水)」
  </span>
  、
  <span
    style="color: rgb(148, 93, 196);"
  >
    「无冕的龙王·北斗(雷)」
  </span>
  、
  <span
    style="color: rgb(236, 73, 35);"
  >
    「智明无邪·烟绯(火)」
  </span>
  中的一个。如果本次祈愿获取的4星物品非本期4星UP角色，下次祈愿获取的4星物品
  <span
    style="color: rgb(201, 63, 35);"
  >
    必定
  </span>
  为本期4星UP角色。当祈愿获取到4星UP物品时，每个本期4星UP角色的获取概率均等。
  <p />
  
  
  <br />
  
  获得4星武器时，会同时获得2个
  <span
    style="color: rgb(189, 105, 50);"
  >
    无主的星辉
  </span>
  作为副产物；获得3星武器时，会同时获得15个
  <span
    style="color: rgb(162, 86, 225);"
  >
    无主的星尘
  </span>
  作为副产物。
  <p />
  
  
  <br />
  
  〓若获得重复角色〓
  <p />
  
  无论通过何种方式（包含但不限于祈愿、商城兑换、系统赠送等）第2~7次获得相同5星角色时，每次将转化为1个
  <span
    style="color: rgb(162, 86, 225);"
  >
    对应角色的命星
  </span>
  和10个
  <span
    style="color: rgb(189, 105, 50);"
  >
    无主的星辉
  </span>
  ；第8次及之后获得，将仅转化为25个
  <span
    style="color: rgb(189, 105, 50);"
  >
    无主的星辉
  </span>
  。
  <p />
  
  无论通过何种方式（包含但不限于祈愿、商城兑换、系统赠送等）第2~7次获得相同4星角色时，每次将转化为1个
  <span
    style="color: rgb(162, 86, 225);"
  >
    对应角色的命星
  </span>
  和2个
  <span
    style="color: rgb(189, 105, 50);"
  >
    无主的星辉
  </span>
  ；第8次及之后获得，将仅转化为5个
  <span
    style="color: rgb(189, 105, 50);"
  >
    无主的星辉
  </span>
  。
  <p />
  
  
  <br />
  
  ※本祈愿属于「角色活动祈愿」，「角色活动祈愿」和「角色活动祈愿-2」的祈愿次数保底完全共享，会一直共同累计在「角色活动祈愿」和「角色活动祈愿-2」中，与其他祈愿的祈愿次数保底相互独立计算，互不影响。
  <p />
</div>
`;
